Upgrade 5.6 (#2042)
openmicroscopy/ome-documentation: omero/developers/python3-migration.rst

Migration from OMERO 5.5 (Python 2) to OMERO 5.6 (Python 3)
===========================================================

> This page serves as a collection of tips the OME team used to

Reviewer: OME team used tips to get through the migration ? 

Maybe better: ``...recommendations, which come up when OME team went through the migration to Python 3.``

> This page serves as a collection of tips the OME team used to

Reviewer: ```suggestion
This page serves as a collection of recommendations, which came up when OME team went through the migration to Python 3.
```

> get through the migration to Python 3. This is not a complete

Reviewer: ```suggestion
This is not a complete
```

guide but may serve as a useful starting point.

For more information, please see a dedicated Python 3 page like
http://python-future.org/.

Futurize
--------

Installing `future` from Python 3 is now required for all OMERO

> Python components. This library comes with a tool, `futurize`

Reviewer: ``...a tool, futurize`` -> ``futurize tool``

> Python components. This library comes with a tool, `futurize`

Reviewer: ```suggestion
Python components. This library comes with `futurize` tool
```

> which will perform many of the basic transformations needed to

Reviewer: ``will perform`` -> ``performs``

> which will perform many of the basic transformations needed to

Reviewer: ```suggestion
which performs many of the basic transformations needed to
```

migrate Python 2 code to Python 3:

::

    futurize -0 your_file.py

Add `-w` to update the file in place.

`print()`
^^^^^^^^^

The most common transformation needed is adding parentheses around
print statements since print is no longer a keyword.

`dict.keys()`
^^^^^^^^^^^^^

The return value from the `keys()` method of dictionaries is of type
`dict_keys` and no longer has methods like `sort()`. Wrap with a call

> to list if you need the previous behavior: `list(my_dict.keys())`.

Reviewer 2: Similar for e.g. ```range(5)```.
Also sort(lambda) and list.map(lambda) have needed updating. Do we want to mention all these?

Author: Happy to have whatever issues added that were problematic for us which we can easily explain to others, but I assume it's more critical to get a version of this PR in early so we have the URLs to link to, then we can iterate on this list.

Strings
-------

> One of if not the major hurdle in upgrading from Python 2 to Python 3 is the shift in

Reviewer: ``One of if not the major hurdle in upgrading from...`` - leave this whole sentence out - the topic is clear from the header of the chapter, otherwise it is not necessary

> One of if not the major hurdle in upgrading from Python 2 to Python 3 is the shift in

Reviewer: ```suggestion

```

> One of if not the major hurdle in upgrading from Python 2 to Python 3 is the shift in

Reviewer: ```suggestion

```

Author: Can you explain why you suggest removing this?

Reviewer: It is a very long sentence which has not much informative value. Makes tedious to read- if this is not the major hurdle, then why do I need to know that that is not the major hurdle ? Obviously it is one of the hurdles, I knew that anyway...

Author: It's a pretty common idiom, but I'm pretty sure we can just say it is the major hurdle. In other words: "this is going to be the problem you run into."

Author: New attempt pushed.

Reviewer: Read the new commit about strings and like it. Ta

> strings. In Python 2, there is a separation between `str` and `unicode`. In Python 3,

Reviewer: ```suggestion
In Python 2, there is a separation between `str` and `unicode`. In Python 3,
```

both of those are like `unicode` (but called `str`) and a new type was introduced:
`bytes`. A good starting places to learn the difference is:

http://python-future.org/compatible_idioms.html?highlight=string#strings-and-bytes

> Complicating the matter further is our use of the `future` library, which tries

Reviewer: ``Complicating the matter further`` -> this is just complicating this text, I would say - leave out

> Complicating the matter further is our use of the `future` library, which tries

Reviewer: ```suggestion
The `future` library tries
```

to help support Python 2 and Python 3 at the same time. Imports including
`past` or `future` should be looked up in the above guide.

What is `str`??
^^^^^^^^^^^^^^^

If nothing special is imported, `str` is the builtin `str` in Py2 (non-unicode) but unicode in Py3.
String literals like "foo" are also of type `str`.

If `unicode_literal` is imported, then "foo" is the same as u"foo" and is `unicode` in Python 2
or just `str` in Python 3.

If `from builtins import str` is imported, then `str` is more like unicode and may fail existing
calls to `isinstance()`.

`isinstance(x, str)`
^^^^^^^^^^^^^^^^^^^^

Since `str` can change its type, this often will not do what you want.
Using `past.builtins.basestring` is generally a good solution, e.g.
`isinstance(x, basestring)`

str(some_variable)
^^^^^^^^^^^^^^^^^^

If you are trying to turn a variable into a string, this may not do what you
want since it might be creating a `unicode`.

This is especially problematic for passing strings to Ice methods, which are

> implemented in the C++ and fail spectacularly if they receive non-string

Reviewer 2: remove "the" C++

objects (like unicode).

`future.utils.native_str` maintains the previous semantics producing builtin `str` objects.
Native str semantics are especially important when working with Ice, e.g.

::

        ctx = {'omero.group': native_str(groupId)}

>         conn.getUpdateService().saveArray(pixels, ctx)

Reviewer 2: But we don't need ```native_str(s)``` for e.g.
```
dataset.name = rstring("test")
```
So, do we only need ```native_str()``` for ctx? I can't think of other examples just now.

Author: Also the `Ice.Properties` methods, at least.

StringIO and open(“file”, “r”)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

`StringIO` and `open()` may need replacing with `BytesIO` and `open("file", "rb")` respectively.
This depends on whether or not your code is expecting a stream to be binary.

Regexes
^^^^^^^

Regexes must start with `r""` for raw to prevent escapes from being misinterpreted (e.g. `\d`).

Numerics
--------

`long` no longer exists. Replace `omero.rtypes.wrap(long_value)` with `omero.rtypes.rlong(long_value)`.

Division with `/` now produces a floating point. For example, `choice * int(percent) / 100` no longer

> produces an int's in Python 3. Use `//`.

Reviewer 2: "an int's" -> "integers" (or ints)?
